ethcore/src/state/mod.rs

```diff
@@ -536,11 +552,86 @@ impl<B: Backend> State<B> {
 	/// Get the storage root of account `a`.
 	pub fn storage_root(&self, a: &Address) -> TrieResult<Option<H256>> {
 		self.ensure_cached(a, RequireCache::None, true,
-			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
+			|a| a.as_ref().and_then(|account| account.storage_root()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the original storage root since last commit of account `a`.
+	pub fn original_storage_root(&self, a: &Address) -> TrieResult<H256> {
+		Ok(self.ensure_cached(a, RequireCache::None, true,
+			|a| a.as_ref().map(|account| account.original_storage_root()))?
+			.unwrap_or(KECCAK_NULL_RLP))
+	}
+
+	/// Get the value of storage at a specific checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
+		enum ReturnKind {
+			/// Use original storage at value at this address.
+			OriginalAt,
+			/// Use the downward checkpoint value.
+			Downward,
+		}
+
+		let (checkpoints_len, kind) = {
+			let checkpoints = self.checkpoints.borrow();
+			let checkpoints_len = checkpoints.len();
+			let checkpoint = match checkpoints.get(checkpoint_index) {
+				Some(checkpoint) => checkpoint,
+				// The checkpoint was not found. Return None.
+				None => return Ok(None),
+			};
+
+			let kind = match checkpoint.get(address) {
+				// The account exists at this checkpoint.
+				Some(Some(AccountEntry { account: Some(ref account), .. })) => {
+					if let Some(value) = account.cached_storage_at(key) {
+						return Ok(Some(value));
+					} else {
+						// This account has checkpoint entry, but the key is not in the entry's cache. We can use
+						// original_storage_at if current account's original storage root is the same as checkpoint
+						// account's original storage root. Otherwise, we need to populate the checkpoint account. Note
+						// that the later case is not possible under current Ethereum consensus rules.
+						if account.base_storage_root() == self.original_storage_root(address)? {
+							ReturnKind::OriginalAt
+						} else {
+							// If account base storage root is different from the original storage root since last
+							// commit, then it can only be created from a new contract, where the base storage root
+							// would always be empty. Note that this branch is actually never called, because
+							// `cached_storage_at` handled this case.
+							return Ok(Some(H256::new()));
```

Comment:
This should indeed be `unreachable!` if we're absolutely certain our code contains no bug. I think we try to make sure the logic is correct as best as we can, but if we made some mistakes elsewhere and it accidentally reached this cause, this reasoning logic provided in this cause is still correct, and we would avoid a client panic.

Reply:
That is what I thought, as I say I prefer to fail than having an unexpected behavior, but I understand that in a production context it is not really possible. Ideally we would have a test build with panics and a production build without. A log alert trace could be a good option for such case.

Reply:
Currently `codecov` shows that this cause is indeed not triggered by any tests. I would actually consider it more like a failsafe (it does not necessarily break consensus) -- if this cause is ever reached, returning `H256::new()` is the only possible correct value.

Added a `warn!` so that if it ever happens, we'll (hopefully) get reports!